src/EditorFeatures/Core/Navigation/AbstractDefinitionLocationService.cs

```diff
@@ -119,24 +130,136 @@ internal abstract class AbstractDefinitionLocationService : IDefinitionLocationS
+                    FindUsagesHelpers.GetDisplayName(method));
+
+                var definitionItem = method.ToNonClassifiedDefinitionItem(solution, includeHiddenLocations: true);
+
+                var referenceItems = new List<SourceReferenceItem>();
+                var classificationOptions = ClassificationOptions.Default with { ClassifyObsoleteSymbols = false };
+                foreach (var documentSpan in documentSpans)
+                {
+                    var classifiedSpans = await ClassifiedSpansAndHighlightSpanFactory.ClassifyAsync(
+                        documentSpan, classifiedSpans: null, classificationOptions, cancellationToken).ConfigureAwait(false);
+
+                    referenceItems.Add(new SourceReferenceItem(
+                        definitionItem, documentSpan, classifiedSpans, SymbolUsageInfo.None, additionalProperties: []));
+                }
+
+                // Multiple document spans this mapped to.  Show them all.
+                return new NavigableLocation(async (options, cancellationToken) =>
+                {
+                    // Can only navigate or present items on UI thread.
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+                    // We have multiple definitions, or we have definitions with multiple locations. Present this to the
+                    // user so they can decide where they want to go to.
+                    //
+                    // We ignore the cancellation token returned by StartSearch as we're in a context where
+                    // we've computed all the results and we're synchronously populating the UI with it.
+                    var (context, _) = _streamingPresenter.StartSearch(title, new StreamingFindUsagesPresenterOptions(SupportsReferences: true));
+                    try
+                    {
+                        await context.OnDefinitionFoundAsync(definitionItem, cancellationToken).ConfigureAwait(false);
+                        await context.OnReferencesFoundAsync(referenceItems.AsAsyncEnumerable(), cancellationToken).ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        await context.OnCompletedAsync(cancellationToken).ConfigureAwait(false);
+                    }
+
+                    return true;
+                });
+            }
         }
+    }
 
-        var definitions = builder.ToImmutable();
+    private static async Task<Dictionary<ImmutableArray<byte>, DocumentId>> ComputeContentHashToDocumentMapAsync(ProjectState projectState, CancellationToken cancellationToken)
+    {
+        var result = new Dictionary<ImmutableArray<byte>, DocumentId>(ByteArrayComparer.Instance);
+
+        foreach (var (documentId, documentState) in projectState.DocumentStates.States)
+        {
+            var text = await documentState.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
>await documentState.GetTextAsync [](http://example.com/codeflow?start=23&length=32)

Is this potentially slow enough to warrant parallelizing the loop?

Reply:
That seems reasonable!

Reply:
actually, this data is now stored on docState and projState.  The first computation will be serial, but after that, it will be cached properly, and updated only when a doc hanges.  so i'm going to keep it serial.